File: web/src/components/artifact/citation-display.ts
/**
 * Map a `OakAnswer` citation `source` to user-facing display text (TestFlight
 * feedback AH1b0N09K/AGGcchFC — internal machinery like `run_sql/natdex_species`
 * or `wiki/<page>` must never leak to non-technical users). The wire `source`
 * string itself is UNCHANGED — it still powers `parseCitationSource` / tappable
 * artifact links; this module only decides what text a user sees.
 *
 * Copy is pinned by the canonical cross-platform copy table (§2) — iOS/Android
 * mirror this behavior exactly. Do not improvise different wording here.
 */

const ENTITY_LABELS: Record<string, string> = {
  pokemon: "Pokémon",
  move: "Move",
  ability: "Ability",
  item: "Item",
  type: "Type",
};

export const MOVEPOOL_LABEL = "Movepool";
export const GAME_DATABASE_LABEL = "Oak's game database";
export const COMMUNITY_WIKI_PREFIX = "Community wiki — ";
export const META_USAGE_GEN9OU_LABEL = "Competitive usage stats (Gen 9 OU)";
export const META_USAGE_LABEL = "Competitive usage stats";

/** "fake-out" -> "Fake Out"; "will-o-wisp" -> "Will O Wisp" (acceptable). */
function titleize(slug: string): string {
  return slug
    .split("-")
    .filter(Boolean)
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
    .join(" ");
}

export function displayCitationSource(source: string): string {
  const slash = source.indexOf("/");
  if (slash <= 0) return source;

  const prefix = source.slice(0, slash).trim();
  const rest = source.slice(slash + 1).trim();
  if (!rest) return source;

  if (prefix === "run_sql") return GAME_DATABASE_LABEL;
  if (prefix === "wiki") return `${COMMUNITY_WIKI_PREFIX}${rest}`;
  if (prefix === "get_meta_usage") {
    return rest === "gen9ou" ? META_USAGE_GEN9OU_LABEL : META_USAGE_LABEL;
  }

  if (prefix === "learnset") {
    // Strip a trailing qualifier like " (gen-9)", same rule as parseCitationSource.
    const paren = rest.indexOf("(");
    const slug = (paren >= 0 ? rest.slice(0, paren) : rest).trim();
    if (!slug) return source;
    return `${MOVEPOOL_LABEL} — ${titleize(slug)}`;
  }

  const entityLabel = ENTITY_LABELS[prefix];
  if (entityLabel) return `${entityLabel} — ${titleize(rest)}`;

  return source;
}

File: ios/OakApp/Features/Chat/AnswerCard/CitationDisplay.swift
import Foundation

/// Maps a `Citation.source` wire string to friendly, non-technical display text
/// (copy-tables.md §2 — mirrors web's `displayCitationSource` exactly). Tap/parse
/// behavior is untouched: this only changes the *visible* text and accessibility
/// label; ``parseCitationSource`` still drives what's tappable and where it opens.
///
/// Internal machinery (`run_sql`, `get_meta_usage`) never surfaces its wire form to
/// the user; entity sources (`pokemon/…`, `move/…`, `learnset/…`, …) render as a
/// kind label + titleized slug, still backed by the same tappable link. Anything
/// unrecognized passes through unchanged.
func displayCitationSource(_ source: String) -> String {
  guard let slashIndex = source.firstIndex(of: "/"), slashIndex != source.startIndex else {
    return source
  }
  let prefix = String(source[source.startIndex..<slashIndex])
    .trimmingCharacters(in: .whitespacesAndNewlines)
  let afterSlash = source.index(after: slashIndex)
  let rest = String(source[afterSlash...])

  switch prefix {
  case "run_sql":
    return "Oak's game database"
  case "wiki":
    return "Community wiki — \(rest)"
  case "get_meta_usage":
    let variant = rest.trimmingCharacters(in: .whitespacesAndNewlines)
    return variant == "gen9ou" ? "Competitive usage stats (Gen 9 OU)" : "Competitive usage stats"
  default:
    break
  }

  guard let parsed = parseCitationSource(source) else { return source }
  return "\(entityLabel(for: prefix)) — \(titleizeSlug(parsed.query))"
}

/// The kind label shown before the slug — keyed on the *raw* wire prefix (not the
/// mapped `EntityKind`) so `learnset/…` reads "Movepool" while `move/…` reads
/// "Move", even though both parse to the same tappable entity kind.
private func entityLabel(for prefix: String) -> String {
  switch prefix {
  case "pokemon": return "Pokémon"
  case "move": return "Move"
  case "ability": return "Ability"
  case "item": return "Item"
  case "type": return "Type"
  case "learnset": return "Movepool"
  default: return prefix
  }
}

/// Splits a hyphenated slug into capitalized words, e.g. `fake-out` → "Fake Out",
/// `will-o-wisp` → "Will O Wisp" (mirrors web's titleize exactly).
private func titleizeSlug(_ slug: String) -> String {
  slug.split(separator: "-")
    .map { word -> String in
      guard let first = word.first else { return String(word) }
      return first.uppercased() + word.dropFirst()
    }
    .joined(separator: " ")
}
